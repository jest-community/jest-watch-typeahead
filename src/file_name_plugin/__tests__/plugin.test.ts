import { KEYS } from 'jest-watcher';
import type { Config } from '@jest/types';
import { jest } from '@jest/globals';

let pluginTester: typeof import('../../test_utils/pluginTester').default = null;
let FileNamePlugin: typeof import('../plugin').default = null;

jest.unstable_mockModule('ansi-escapes', () => ({
  default: {
    clearScreen: '[MOCK - clearScreen]',
    cursorDown: (count = 1) => `[MOCK - cursorDown(${count})]`,
    cursorLeft: '[MOCK - cursorLeft]',
    cursorHide: '[MOCK - cursorHide]',
    cursorRestorePosition: '[MOCK - cursorRestorePosition]',
    cursorSavePosition: '[MOCK - cursorSavePosition]',
    cursorShow: '[MOCK - cursorShow]',
    cursorTo: (x, y) => `[MOCK - cursorTo(${x}, ${y})]`,
  },
}));

jest.doMock('ansi-escapes', () => ({
  clearScreen: '[MOCK - clearScreen]',
  cursorDown: (count = 1) => `[MOCK - cursorDown(${count})]`,
  cursorLeft: '[MOCK - cursorLeft]',
  cursorHide: '[MOCK - cursorHide]',
  cursorRestorePosition: '[MOCK - cursorRestorePosition]',
  cursorSavePosition: '[MOCK - cursorSavePosition]',
  cursorShow: '[MOCK - cursorShow]',
  cursorTo: (x, y) => `[MOCK - cursorTo(${x}, ${y})]`,
}));

const projects: Config.ProjectConfig[] = [
  {
    config: {
      rootDir: '/project',
    },
    testPaths: ['/project/src/foo.js', '/project/src/file-1.js'],
  },
  {
    config: {
      rootDir: '/project',
    },
    testPaths: ['/project/src/bar.js', '/project/src/file-2.js'],
  },
];

beforeAll(async () => {
  FileNamePlugin = (await import('../plugin')).default;
  pluginTester = (await import('../../test_utils/pluginTester')).default;
});

it('shows the correct initial state', async () => {
  const { stdout, hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(FileNamePlugin);

  hookEmitter.onFileChange({ projects });
  const runPromise = plugin.run({}, updateConfigAndRun);
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();
  type(KEYS.ENTER);

  await runPromise;
});

it('can use arrows to select a specific file', async () => {
  const { stdout, hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(FileNamePlugin);

  hookEmitter.onFileChange({ projects });
  const runPromise = plugin.run({}, updateConfigAndRun);
  stdout.write.mockReset();
  type('f', 'i', KEYS.ARROW_DOWN, KEYS.ENTER);
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();

  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testPathPattern: 'src/file-1\\.js',
  });
});

it('can select a specific file that includes a regexp special character', async () => {
  const { hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(FileNamePlugin);

  hookEmitter.onFileChange({
    projects: [
      {
        config: {
          rootDir: '/project',
        },
        testPaths: ['/project/src/file_(xyz).js'],
      },
    ],
  });
  const runPromise = plugin.run({}, updateConfigAndRun);

  type('x', 'y', 'z', KEYS.ARROW_DOWN, KEYS.ENTER);

  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testPathPattern: 'src/file_\\(xyz\\)\\.js',
  });
});

it('can select a pattern that matches multiple files', async () => {
  const { stdout, hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(FileNamePlugin);

  hookEmitter.onFileChange({ projects });
  const runPromise = plugin.run({}, updateConfigAndRun);
  stdout.write.mockReset();
  type('f', 'i', KEYS.ENTER);
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();

  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testPathPattern: 'f.*i',
  });
});

it('can configure the key and prompt', async () => {
  const { plugin } = pluginTester(FileNamePlugin, {
    config: {
      key: 'l',
      prompt: 'have a custom prompt',
    },
  });

  expect(plugin.getUsageInfo()).toEqual({
    key: 'l',
    prompt: 'have a custom prompt',
  });
});

it('file matching is case insensitive', async () => {
  const { stdout, hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(FileNamePlugin);

  hookEmitter.onFileChange({ projects });
  const runPromise = plugin.run({}, updateConfigAndRun);
  type('f');
  stdout.write.mockReset();
  type('I');
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();
  type(KEYS.ENTER);
  await runPromise;
});

it("selected file doesn't include trimming dots", async () => {
  const { hookEmitter, updateConfigAndRun, plugin, type } = pluginTester(
    FileNamePlugin,
    {
      stdout: { columns: 40 },
    },
  );

  hookEmitter.onFileChange({
    projects: [
      {
        config: {
          rootDir: '/project',
        },
        testPaths: ['/project/src/long_name_gonna_need_trimming.js'],
      },
    ],
  });
  const runPromise = plugin.run({}, updateConfigAndRun);

  type('t', 'r', 'i', 'm', 'm', KEYS.ARROW_DOWN, KEYS.ENTER);
  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testPathPattern: 'ong_name_gonna_need_trimming\\.js',
  });
});

it('can fuzzy search', async () => {
  const { stdout, hookEmitter, updateConfigAndRun, plugin, type } =
    pluginTester(FileNamePlugin);

  hookEmitter.onFileChange({ projects });
  const runPromise = plugin.run({}, updateConfigAndRun);
  stdout.write.mockReset();
  type('f', '1', KEYS.ARROW_DOWN, KEYS.ENTER);
  expect(stdout.write.mock.calls.join('\n')).toMatchSnapshot();

  await runPromise;

  expect(updateConfigAndRun).toHaveBeenCalledWith({
    mode: 'watch',
    testPathPattern: 'src/file-1\\.js',
  });
});
