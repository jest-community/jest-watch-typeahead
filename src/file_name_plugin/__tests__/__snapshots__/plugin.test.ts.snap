// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can fuzzy search 1`] = `

[MOCK - cursorLeft]

 pattern › f
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Pattern matches 3 files

 › src/foo.js

 › src/file-1.js

 › src/file-2.js
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]

[MOCK - cursorLeft]

 pattern › f1
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Pattern matches 1 file

 › src/file-1.js
[MOCK - cursorTo(13, 5)]
[MOCK - cursorRestorePosition]

[MOCK - cursorLeft]

 pattern › f1
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Pattern matches 1 file

 › src/file-1.js
[MOCK - cursorTo(13, 5)]
[MOCK - cursorRestorePosition]
`;

exports[`can select a pattern that matches multiple files 1`] = `

[MOCK - cursorLeft]

 pattern › f
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Pattern matches 3 files

 › src/foo.js

 › src/file-1.js

 › src/file-2.js
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]

[MOCK - cursorLeft]

 pattern › fi
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Pattern matches 2 files

 › src/file-1.js

 › src/file-2.js
[MOCK - cursorTo(13, 5)]
[MOCK - cursorRestorePosition]
`;

exports[`can use arrows to select a specific file 1`] = `

[MOCK - cursorLeft]

 pattern › f
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Pattern matches 3 files

 › src/foo.js

 › src/file-1.js

 › src/file-2.js
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]

[MOCK - cursorLeft]

 pattern › fi
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Pattern matches 2 files

 › src/file-1.js

 › src/file-2.js
[MOCK - cursorTo(13, 5)]
[MOCK - cursorRestorePosition]

[MOCK - cursorLeft]

 pattern › fi
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Pattern matches 2 files

 › src/file-1.js

 › src/file-2.js
[MOCK - cursorTo(13, 5)]
[MOCK - cursorRestorePosition]
`;

exports[`file matching is case insensitive 1`] = `

[MOCK - cursorLeft]

 pattern › fI
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Pattern matches 2 files

 › src/file-1.js

 › src/file-2.js
[MOCK - cursorTo(13, 5)]
[MOCK - cursorRestorePosition]
`;

exports[`shows the correct initial state 1`] = `
[MOCK - cursorHide]
[MOCK - clear]

Pattern Mode Usage
 › Press Esc to exit pattern mode.
 › Press Enter to filter by a filenames regex pattern.


[MOCK - cursorShow]

[MOCK - cursorLeft]

 pattern › 
[MOCK - cursorSavePosition]
[MOCK - cursorLeft]


 Start typing to filter by a filename regex pattern.
[MOCK - cursorTo(11, 5)]
[MOCK - cursorRestorePosition]
`;
